maltoolbox/model.py

```diff
@@ -2,58 +2,61 @@
 MAL-Toolbox Model Module
 """
 
+from __future__ import annotations
+from dataclasses import dataclass, field
 import json
 import logging
+from typing import Any, Dict, List, Optional, Set, Tuple
 
-from dataclasses import dataclass
+from maltoolbox.language import LanguageClassesFactory
 from maltoolbox.file_utils import (
     load_dict_from_json_file, load_dict_from_yaml_file,
     save_dict_to_file
 )
 
-from .__init__ import __version__
+from maltoolbox import __version__
 
 logger = logging.getLogger(__name__)
 
 @dataclass
 class AttackerAttachment:
     """Used to attach attackers to attack step entrypoints of assets"""
-    id: int = None
-    name: str = None
-    entry_points: list[tuple] = None
+    id: Optional[int] = None
+    name: Optional[str] = None
+    entry_points: list[tuple] = field(default_factory=lambda: [])
 
 class Model():
     """An implementation of a MAL language with assets and associations"""
-    next_id = 0
+    next_id: int = 0
 
     def __repr__(self) -> str:
         return f'Model {self.name}'
 
     def __init__(
             self,
-            name,
-            lang_classes_factory,
-            mt_version = __version__
+            name: str,
+            lang_classes_factory: LanguageClassesFactory,
+            mt_version: str = __version__
         ):
 
         self.name = name
-        self.assets = []
-        self.associations = []
-        self.attackers = []
-        self.lang_classes_factory = lang_classes_factory
-        self.maltoolbox_version = mt_version
+        self.assets: List = []
+        self.associations: List = []
+        self.attackers: List[AttackerAttachment] = []
+        self.lang_classes_factory: LanguageClassesFactory = lang_classes_factory
+        self.maltoolbox_version: str = mt_version
 
         # Below sets used to check for duplicate names or ids,
         # better for optimization than iterating over all assets
-        self.asset_ids = set()
-        self.asset_names = set()
+        self.asset_ids: Set[int] = set()
+        self.asset_names: Set[str] = set()
 
     def add_asset(
             self,
-            asset,
-            asset_id: int = None,
-            allow_duplicate_names = True
-        ):
+            asset: Any,
```

Comment:
Isn't there a more specific choice here? What is the common class of assets? Sth from json schema objects maybe? Same for associations further down.

Reply:
There is a class that they share (python_jsonschema_objects.classbuilder.ProtocolBase) but it would only be confusing to use I think

Reply:
We can create a type alias that points to weird thing but has a nice name. I think it will be nice for development to hint where this is supposed to actually come from.

Reply:
Good solution for it, will look into it.

Reply:
I did not manage to do this without mypy flipping out (about all the fields id,name,extras not existing for the type + other stuff), so I will keep it as Any for now. If someone knows a solution let me know.

Reply:
There is a solution, a bit hacky, but I think it is worth having together with a couple comment lines to explain to the future users what is going on:

```
if TYPE_CHECKING:
    from python_jsonschema_objects.classbuilder import Protocol as RealProtocolBase
    from typing import Any, TypeAlias

    class ProtocolBase(RealProtocolBase):
        def __getattr__(self, name: str) -> Any:
            return Any

    SchemaGeneratedClass: TypeAlias = ProtocolBase
```

Reply:
I am not sure about if it is worth it, it's so much code just for the type hint

Reply:
I would add it as it is one of the confusing parts of the code. An alternative could be to rename `asset`. But personally, I could have used some assistance when I read the code the first time, what what this asset, where did it come from, etc. In any case, whatever you two decide.

Reply:
What about this?

```
    from python_jsonschema_objects.classbuilder import ProtocolBase

    # Class below created as Type alias for objects of classes
    # generated by python_jsonschema_objects
    class SchemaGeneratedClass(ProtocolBase): ...

```

We can then use `SchemaGeneratedClass` as type for assets and associations
Hacky, but less code than the one you suggested.

Reply:
Have you tested this? I am not sure mypy will not complain about id, name etc being invalid fields. The trick in my example was the presence of `__getattr__` that informed mypy about the dynamic nature of the object's fields.

Reply:
Yes, somehow it works :o

Reply:
Pushed now so you can see, should pass the pipelines (with mypy)

Reply:
Check my commit, is it possible that you messed sth when you tried with the type alias?

Reply:
That looks great, definitely possible :)